CONTRIBUTING.md

```diff
@@ -161,6 +161,12 @@ Pull requests are appreciated. If your PR includes any changes that would impact
 
 Before we can merge your PR, it must pass automated continuous integration checks in each of the supported environments, as well as a check to ensure that code coverage has not decreased significantly.
 
+## Branch name conventions
+
+Each branch should start with the major version of the release the branch is targeting. For example:
```

Comment:
Or here's another wild option for picking a base version:
What if we define multiple `type: approval` jobs, each one for a specific base version. If we need to compile an API diff for a specific version - we just "approve" the corresponding flow which configures the build steps with predefined base tags. Thus it is possible to maintain compatibility for old versions and well as append new releases easily.

Reply:
Thanks for good feedback. Below is my answers:

> Is it mandatory now to follow this pattern for each branch? 

- Good question. The answer is yes, and we can provide a step which will verify the branch name as an improvement. 

> I think that naming decision deserves a discussion

- The naming is up to discussion. Feel free to propose your variant with a motivation note. I picked to embed the version because it provides a way to see all v2 PRs in the first level (GUI git clients usually represent such branches as a path with folders). It is a more logical variant, in my opinion, but I don't have strong preferences. We can discuss the naming. 

> Also, I see that docs suggests trying v1

- [ ] v1 should be excluded from the docs for now because it is not a supported configuration. Thanks for the suggestion. 

> What if we define multiple type: approval jobs

- Having approval jobs pauses all the workflows until you explicitly accept it. That's why I rejected this variant. 

> Could we parse branch name to extract version number automatically

- I wanted to use automatic mapping from a target branch to base API commit but haven't found a way to do this due to CircleCI configuration limitations:
  - Cache keys can't have arbitrary data, and you can only use a limited set of env variables that are not suitable for our needs (see https://circleci.com/docs/2.0/configuration-reference/#savecache). 
  - There is no easy way to get the target branch of the PR. Even though this issue can be solved, the first limitation still blocks better implementation. 
If you have a suggestion how to implement this dynamic branching, let me know. 
  
> append new releases easily.

- Currently, we have only one supported version for the foreseeable future, so I don't think it makes sense to spend more time on optimizing adding new versions while there is no one on the horizon. 

> Thus it is possible to maintain compatibility for old versions

- We are following semantic versioning, and according to it, you shall introduce a major release when you have a breaking change. So, I don't think it makes sense to support such configuration in API diff workflow. Even if we need (very unlikely, in my opinion), we can add a new convention to the branch name or rethink this workflow.

Reply:
Does circle ci have any ability to access both _target_ and _base_ branches in the PR? We just need to make sure that there is no changes between this two.

Reply:
- There is no built-in functionality to access target branch.
- I haven't found a way to cache target branch API diffs if I get this target branch by other means. Cache keys are ver limiting.

Reply:
> Also, I see that docs suggests trying v1

Well, it turns out it is ok to keep the mention of v1 in CONTRIBUTING guide. It is just a convention. Whether CI will pick this convention isn't important here.

Reply:
> There is no built-in functionality to access target branch.

I’m surprised CircleCI doesn’t provide this functionality out of the box like other CI services do. I wonder if [this workaround](https://discuss.circleci.com/t/create-a-circle-target-branch-envar/10022/4) would be suitable.

Reply:
> Each branch should start with the major version of the release the branch is targeting.

This means if we ever decide to land a PR in a different major version, we can’t just rebase and retarget – we’d have to close the PR and open a new one. If it isn’t feasible to get the target branch, would it be feasible to get the milestone?

Reply:
> I’m surprised CircleCI doesn’t provide this functionality out of the box like other CI services do. I wonder if this workaround would be suitable.

>> Cache keys can't have arbitrary data, and you can only use a limited set of env variables that are not suitable for our needs (see https://circleci.com/docs/2.0/configuration-reference/#savecache).

Reply:
> This means if we ever decide to land a PR in a different major version, we can’t just rebase and retarget – we’d have to close the PR and open a new one.

Is it a problem to solve? What's bad about closing the PR? We already do this time to time. Just link the original PR if it has important conversation. 

>  If it isn’t feasible to get the target branch, would it be feasible to get the milestone?

The problem not that we can't get target branch, the problem is that we can't load cache for target branch, because we can't form the correct key. If we don't cache API results, it can be done.

Reply:
> Is it a problem to solve? What's bad about closing the PR? We already do this time to time. Just link the original PR if it has important conversation.

We actually move PRs from milestone to milestone quite often.

If the problem is specific to caching, what about a solution that requires us to do something explicit to cause the check to run? For example, I’ve seen CI integrations that require someone to post a comment on the PR with a special keyword that triggers a run of some expensive or noncritical check. I realize my suggestion would be scope-creep, but maybe we could live with not caching the results in the meantime. We did go without caching at all in this repository until #1632; the main downside was longer build times.

Reply:
I understand your concerns. I'm trying different approach which will be even better in my opinion.

Reply:
Fixed